From: Antoine Musso <hashar@free.fr>
Date: Mon, 18 Jul 2016 12:11:13 +0200
Subject: Gerrit trailing delay is now configurable

Zuul handling of events is always behind Gerrit by 10 seconds due to
change dependencies tracking being off immediately after a patchset has
been uploaded.

Make it a configuration option to easily fine tune the delay. On some
third parties setup the delay can be significantly reduced.

GerritEventConnector() delay moved to a class parameter, default to 10.

GerritConnection() now recognizes the optional configuration setting
'event_delay' in zuul.conf representing the amount of seconds to delay.
Default to 10 as well.

Change the way test/base.py set the GerritEventConnector delay by
injecting the 'event_delay' setting when the test suite load the
configuration.

Fix an unrelated typo in GerritConnection debug log.

Update documentation inspired by James E. Blair code comment and commit
message summaries that introduced the delay:

    5241b88 Delay Gerrit events by 5s
    490f4aa Increase the Gerrit trailing delay

Change-Id: I0179d5bac6e0b3313e44e850823385345d28cb9d
---
 doc/source/connections.rst | 12 ++++++++++++
 etc/zuul.conf-sample       |  1 +
 tests/base.py              |  2 +-
 zuul/connection/gerrit.py  | 14 +++++++++++---
 4 files changed, 25 insertions(+), 4 deletions(-)

diff --git a/doc/source/connections.rst b/doc/source/connections.rst
index f0820a6..c1753da 100644
--- a/doc/source/connections.rst
+++ b/doc/source/connections.rst
@@ -38,6 +38,18 @@ Create a connection with gerrit.
   Path to SSH key to use when logging into above server.
   ``sshkey=/home/zuul/.ssh/id_rsa``
 
+**event_delay** (optional)
+
+  When querying a change immediately after a patchset upload, Gerrit may
+  return incorrect data about dependent changes. In order to avoid this,
+  the events are not delivered to Zuul until a constant number of
+  seconds has passed.
+
+  Note that if we receive several events in succession, we will only
+  need to delay for the first event.
+
+  Default: ``10`` (seconds)
+
 
 Gerrit Configuration
 ~~~~~~~~~~~~~~~~~~~~
diff --git a/etc/zuul.conf-sample b/etc/zuul.conf-sample
index 21c1317..77ed59e 100644
--- a/etc/zuul.conf-sample
+++ b/etc/zuul.conf-sample
@@ -32,6 +32,7 @@ server=review.example.com
 ;baseurl=https://review.example.com/r
 user=jenkins
 sshkey=/home/jenkins/.ssh/id_rsa
+event_delay=10
 
 [connection smtp]
 driver=smtp
diff --git a/tests/base.py b/tests/base.py
index 4f41be9..541a8c1 100755
--- a/tests/base.py
+++ b/tests/base.py
@@ -927,7 +927,6 @@ class ZuulTestCase(BaseTestCase):
 
         zuul.source.gerrit.GerritSource.replication_timeout = 1.5
         zuul.source.gerrit.GerritSource.replication_retry_interval = 0.5
-        zuul.connection.gerrit.GerritEventConnector.delay = 0.0
 
         self.sched = zuul.scheduler.Scheduler(self.config)
 
@@ -1018,6 +1017,7 @@ class ZuulTestCase(BaseTestCase):
                         Queue.Queue()
                     self.event_queues.append(
                         self.gerrit_queues_dbs[con_config['server']])
+                con_config['event_delay'] = 0.0
                 self.connections[con_name] = FakeGerritConnection(
                     con_name, con_config,
                     changes_db=self.gerrit_changes_dbs[con_config['server']],
diff --git a/zuul/connection/gerrit.py b/zuul/connection/gerrit.py
index a1854f4..231ae30 100644
--- a/zuul/connection/gerrit.py
+++ b/zuul/connection/gerrit.py
@@ -32,13 +32,17 @@ class GerritEventConnector(threading.Thread):
     """Move events from Gerrit to the scheduler."""
 
     log = logging.getLogger("zuul.GerritEventConnector")
+<<<<<<< HEAD
     delay = 5.0
+=======
+>>>>>>> 12d23bf... Gerrit trailing delay is now configurable
 
-    def __init__(self, connection):
+    def __init__(self, connection, delay=10):
         super(GerritEventConnector, self).__init__()
         self.daemon = True
         self.connection = connection
         self._stopped = False
+        self.delay = delay
 
     def stop(self):
         self._stopped = True
@@ -54,6 +58,8 @@ class GerritEventConnector(threading.Thread):
         # that if we receive several events in succession, we will
         # only need to delay for the first event.  In essence, Zuul
         # should always be a constant number of seconds behind Gerrit.
+        #
+        # Can be configured via the Gerrit driver setting 'event_delay'.
         now = time.time()
         time.sleep(max((ts + self.delay) - now, 0.0))
         event = TriggerEvent()
@@ -226,6 +232,7 @@ class GerritConnection(BaseConnection):
         self.keyfile = self.connection_config.get('sshkey', None)
         self.watcher_thread = None
         self.event_queue = None
+        self.event_delay = int(self.connection_config.get('event_delay', 10))
         self.client = None
 
         self.baseurl = self.connection_config.get('baseurl',
@@ -440,7 +447,7 @@ class GerritConnection(BaseConnection):
         return url
 
     def onLoad(self):
-        self.log.debug("Starting Gerrit Conncetion/Watchers")
+        self.log.debug("Starting Gerrit Connection/Watchers")
         self._start_watcher_thread()
         self._start_event_connector()
 
@@ -470,7 +477,8 @@ class GerritConnection(BaseConnection):
             self.gerrit_event_connector.join()
 
     def _start_event_connector(self):
-        self.gerrit_event_connector = GerritEventConnector(self)
+        self.gerrit_event_connector = GerritEventConnector(
+            self, delay=self.event_delay)
         self.gerrit_event_connector.start()
 
 
